### ISSUE TITLE
Keep comments associated with a condition at the condition's indentation level

### ISSUE THREAD

**Issue by Loilo** [prettier/plugin-php#1000]
> **Note:** This very same issue has been open for a pretty long time over at [the Prettier repo](https://github.com/prettier/prettier/issues/807). It's gotten positive signals from maintainers, but nobody has taken on it yet.

**@prettier/plugin-php v0.10.2**
[Playground link](https://prettier-php-playground-czxafnpemd.now.sh/#N4IgDgTgpgLjCWUIHkwIPYDsDOIBcok8mMA6vACYwAW+AHAAwA0IMAhgEblW14AsLAK7YoAFU648AMzYAbES2zEA5rKgBFQehhR8M+VBYwIbeLJUBhdAFtrbfCExZdLDiYDGUAMowAnmocwbAgAWgAmEBZoAEdBeGgABRNlOz05BRBiEQgYJLYU+2l0qABfFigKeBh0FDR4LElQNmwYNIMyzMwwQVa8EAAeAH4wajAAHUwJgHopgAIAEXRZ7BtYahUJ+ClZgAp3LEqMTABKWeAJ2cvZinQAfRXrNZUd44BuCem5gFEADzBZUyYFSzADu1F813QwLYThoSFmNA2mBKsygBjOFyuN1uMO01CQt0RmGUL3eyJAJSAA)

**Input:**
```php
<?php

// Do something
if (condition) {
    do_something();

// Explaining why doing another thing
} else {
    do_another_thing();
}
```

**Output:**
```php
<?php

// Do something
if (condition) {
    do_something();

    // Explaining why doing another thing
} else {
    do_another_thing();
}

```

**Expected behavior:**
```php
<?php

// Do something
if (condition) {
    do_something();

// Explaining why doing another thing
} else {
    do_another_thing();
}

```

**Reasoning:**
Comments that can clearly be associated with a condition should stick with that condition, indentation-wise.

Instead, the comment is (incorrectly) recognized as part of the `if` block and indented with it.

**Resolving:**
I'm going to list the characteristics to recognize comments that should be aligned with their associated conditions. These characteristics are used in conjunction, so *all* of them need to apply to align a comment with its condition.

However, there's one precondition which I'll send ahead: For the following characteristics, multiple successive single-line comments with the same indentation level and with no empty lines between them, are *one* comment.

It's easier to grasp in code:
```php
// In regard of the following characteristics,
// this is actually a single comment.
```

So, the characteristics for a comment to be intended with its condition are the following:

* ### The comment ends on the last line before a conditional keyword after a brace.
  **✅ Applies to:**
  ```php
  if (...) {
      // ...

  // Last line before the conditional keyword
  } else {
  }
  ```

  **❌ Does not apply to:**
  ```php
  if (...) {
      // ...

  // The conditional keyword is not on next line
  }
  else {
  }
  ```

* ### The comment has the same indentation level as the conditional keyword.
  **✅ Applies to:**
  ```php
  if (...) {
      // ...

  // Same level
  } else {
  }
  ```

  **❌ Does not apply to:**
  ```php
  if (...) {
      // ...

      // Not the same level
  } else {
  }
  ```

* ### The associated conditional keyword continues an existing condition.
  **✅ Applies to:**
  ```php
  if (...) {
      // ...

  // The `else` refers to the `if` block above it
  } else {
  }
  ```

  **❌ Does not apply to:**
  ```php
  // The `if` opens the first branch of a condition
  if (...) {
      // ...
  }
  ```
  ```php
  if (...) {
      // ...

  // Nobody would ever do that, but it needs to be specified
  } if (...) {

  }
  ```

* ### The continued condition has a body wrapped in braces.
  **✅ Applies to:**
  ```php
  if (...) {
      // ...

  // The `if` body is wrapped in braces
  } else {
  }
  ```

  **❌ Does not apply to:**
  ```php
  if (...)
      // ...

  // The `if` body is not wrapped in braces
  else {
  }
  ```

* ### The surrounding block contains code other than the comment.
  **✅ Applies to:**
  ```php
  if (...) {
      $foo = 'bar';

  // Comment
  } else {
  }
  ```
  ```php
  if (...) {
      // Comment

  // Comment
  } else {
  }
  ```
  ```php
  if (...) {
  // This code example is probably debateable.

  // Comment
  } else {
  }
  ```

  **❌ Does not apply to:**
  ```php
  if (...) {
  // Comment
  } else {
  }
  ```
  ```php
  if (...) {
  // Does not apply because this is
  // a single comment as per precondition
  } else {
  }
  ```

**Conclusion:** What do you think?

**Comment by evilebottnawi** [comment]
I agree, note about this exists in https://github.com/prettier/plugin-php/issues/829

**Comment by evilebottnawi** [comment]
We already support `try`/`catch`/`finally`

**Comment by Loilo** [comment]
Yeah, the plugin already formats comments outside the condition body correctly:

```php
if (...) {
}
// Comment
else {
}
```

That's why I explicitely excluded those in the listed characteristics. :)

**Comment by Loilo** [comment]
Oh, also I just realized: issue #1000! 🎉
